crates/types/src/consensus.rs

```diff
@@ -433,11 +476,43 @@ impl<TYPES: NodeType> Consensus<TYPES> {
         Ok(())
     }
 
+    /// Update the last actioned view internally for votes and proposals
+    ///
+    /// Returns true if the action is for a newer view than the last action of that type
+    pub fn update_action(&mut self, action: HotShotAction, view: TYPES::Time) -> bool {
+        let old_view = match action {
+            HotShotAction::Vote => &mut self.last_actions.voted,
+            HotShotAction::Propose => &mut self.last_actions.proposed,
+            HotShotAction::DaPropose => &mut self.last_actions.da_proposed,
+            HotShotAction::DaVote => {
+                if view > self.last_actions.da_vote {
+                    self.last_actions.da_vote = view;
+                }
+                // TODO Add logic to prevent double voting.  For now the simple check if
+                // the last voted view is less than the view we are trying to vote doesn't work
+                // becuase the leader of view n + 1 may propose to the DA (and we would vote)
+                // before the leader of view n.
+                return true;
+            }
+            _ => return true,
+        };
+        if view > *old_view {
+            *old_view = view;
+            return true;
+        }
+        false
+    }
+
+    /// reset last actions to genesis so we can resend events in tests
+    pub fn reset_actions(&mut self) {
```

Comment:
is there a way to gate this to tests only? or do we want to potentially use it elsewhere in the future

Reply:
I tried to gate this for test but it causes a whole bunch of issues because lots of the byzantine code isn't gated for test etc.  I'll follow up with the gating because we should gate this